psi4/src/psi4/libscf_solver/sad.cc

```diff
@@ -358,61 +358,70 @@ void SADGuess::get_uhf_atomic_density(std::shared_ptr<BasisSet> bas, std::shared
     SharedMatrix Fa(mat.create_matrix("Fa"));
     SharedMatrix Fb(mat.create_matrix("Fb"));
 
-    // Factional occupation
+    // Fractional occupation
     SharedVector occ_a, occ_b;
-    if (options_.get_bool("SAD_FRAC_OCC")) {
-        int nfzc = 0, nact = 0;
-        if (Z <= 2) {
-            nfzc = 0;
-            nact = 1;
-        } else if (Z <= 4) {
-            nfzc = 1;
-            nact = 1;
-        } else if (Z <= 10) {
-            nfzc = 2;
-            nact = 3;
-        } else if (Z <= 18) {
-            nfzc = 5;
-            nact = 4;
-        } else if (Z <= 36) {
-            nfzc = 9;
-            nact = 9;
-        } else if (Z <= 54) {
-            nfzc = 18;
-            nact = 9;
-        } else if (Z <= 54) {
-            nfzc = 18;
-            nact = 9;
-        } else if (Z <= 86) {
-            nfzc = 27;
-            nact = 16;
+    // Number of partially or fully occupied orbitals
+    int nocc_a, nocc_b;
+    if (options_.get_bool("SAD_FRAC_OCC") || options_.get_bool("SAD_FRAC_SR_OCC")) {
+        // The density is spread over the whole of the possible valence shell.
+        // Only the noble gas core is doubly occupied
+        static const std::vector<int> magic_values = {0, 2, 10, 18, 36, 54, 86, 118};
+        // Shell index
+        size_t imagic = 0;
+        while(Z>magic_values[imagic+1]) {
+          imagic++;
+          if(imagic+1>=magic_values.size())
```

Comment:
A couple of trivial comments about this part.  a) because line 380 uses the `imagic+1` element, i think this needs to test `imagic+2 >= ...` b) the `std::upper_bound` algorithm implements this kind of search c) if the intention is to freeze the full noble gas part, shouldn't the test on line 371 be greater than or equal to (which would be equivalent to the `std::lower_bound` result)?

Reply:
Right, I can use algorithms now that I changed the code to using C++ vector. Done.